python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,107 +871,120 @@ class PyInferenceRequest
   {
   }
 
-
-  // Use internally when interacting with C APIs that takes ownership,
-  // this function will also release the ownership of the callback resource
-  // because once the ownership is transferred, the callback resource
-  // will be accessed in the callback pipeline and should not be tied to the
-  // PyWrapper's lifecycle. The callback resource will be released in the
-  // Triton C callback wrapper.
-  struct TRITONSERVER_InferenceRequest* Release()
+  void SetReleaseCallback(const std::shared_ptr<PyInferenceRequest>& request)
   {
-    // Note that Release() doesn't change ownership as the
-    // same PyInferenceRequest will be passed along the life cycle.
-    allocator_callback_resource_.release();
-    response_callback_resource_.release();
-    return triton_object_;
+    // The request wrapper needs to be kept alive until the release callback is
+    // invoked, so the Triton request can be deleted after the core/backend is
+    // done with it.
+    std::unique_ptr<std::shared_ptr<PyInferenceRequest>> request_ptr(
+        new std::shared_ptr<PyInferenceRequest>(request));
+    ThrowIfError(TRITONSERVER_InferenceRequestSetReleaseCallback(
+        triton_object_, ReleaseCallback,
+        reinterpret_cast<void*>(
+            request_ptr.release()) /* request_release_userp */));
+  }
+  static void ReleaseCallback(
+      struct TRITONSERVER_InferenceRequest* request, const uint32_t flags,
+      void* userp)
+  {
+    std::unique_ptr<std::shared_ptr<PyInferenceRequest>> request_ptr(
+        reinterpret_cast<std::shared_ptr<PyInferenceRequest>*>(userp));
   }
 
-  struct CallbackResource {
-    CallbackResource(ReleaseFn r, const py::object& uo)
-        : release_fn(r), user_object(uo)
+  struct ResponseAllocatorDeleter {
+    void operator()(struct TRITONSERVER_ResponseAllocator* allocator) const
     {
-    }
-    ReleaseFn release_fn;
-    py::object user_object;
-    // Unsafe handling to ensure the same PyInferenceRequest object
-    // goes through the request release cycle. This is due to
-    // a 'keep_alive' relationship is built between 'PyInferenceRequest'
-    // and 'PyServer': a request is associated with a server and the server
-    // should be kept alive until all associated requests is properly released.
-    // And here we exploit the 'keep_alive' utility in PyBind to guarantee so.
-    // See PyServer::InferAsync on how this field is set to avoid potential
-    // circular inclusion.
-    std::shared_ptr<PyInferenceRequest> request;
+      if (allocator != nullptr) {
+        ThrowIfError(TRITONSERVER_ResponseAllocatorDelete(allocator));
+      }
+    };
   };
-
-
-  void SetReleaseCallback(ReleaseFn release, const py::object& user_object)
+  void SetResponseAllocator()
+  {
+    // The response allocator is shared among all instances of this class, so it
+    // needs to be initialized once and once only.
+    // TODO: Why 'numpy.ones(2**27)' gets stuck when the response allocator is
```

Comment:
seems very odd like something is interacting between numpy and this allocator - and that doesn't seem to make sense to me ....

Reply:
my observation: the issue does not happen as long as we are not allocating memory during the static initialization phase

cc @GuanLuo if you have any thought on this?